---
title: SY-179下潜样品v6-v8扩增子分析
tags: 扩增子,v6-v8,MiSeq PE300,下潜,114 Samples
renderNumberedHeading: true
grammar_cjkRuby: true
---


# 准备输入文件列表
- 双端数据
```bash
$cd ~/Datas/16S_nanhai/Analysis
$echo -e "sample-id\tforward-absolute-filepath\treverse-absolute-filepath\n" > manifest.tsv
$realpath PATHtoReads/*.R1.fq.gz >> manifest.tsv
```
__注：__正则表达式提取IDs至首列，

__匹配：__`((.+_)(.+).R1.fq.gz)`

__替换为：__`$3\t$1\t$2$3.R2.fq.gz`

__文件内容如下所示：__

```
sample-id	forward-absolute-filepath	reverse-absolute-filepath
F01	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1859_F01_clean.R1.fq.gz	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1859_F01_clean.R2.fq.gz
F02	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1860_F02_clean.R1.fq.gz	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1860_F02_clean.R2.fq.gz
F03	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1861_F03_clean.R1.fq.gz	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1861_F03_clean.R2.fq.gz
F04	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1862_F04_clean.R1.fq.gz	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1862_F04_clean.R2.fq.gz
F05	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1863_F05_clean.R1.fq.gz	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1863_F05_clean.R2.fq.gz
F06	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1864_F06_clean.R1.fq.gz	/new_data/hualin/sea_mountain/v6v8/reads/S20201230-1864_F06_clean.R2.fq.gz
```

# 导入数据
real	2m43.071s
```
time qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' --input-path manifest.tsv --output-path paired-end-demux.qza --input-format PairedEndFastqManifestPhred33V2
```

# 检查样本的序列和测序深度
real	3m38.199s
```
time qiime demux summarize --i-data paired-end-demux.qza --o-visualization demux-summary-1.qzv
```
# 序列质量控制和特征表
real	123m52.192s
```
time qiime dada2 denoise-paired --p-trim-left-f 20 --p-trim-left-r 15 --p-trunc-len-f 296 --p-trunc-len-r 286 --p-n-threads 20 --i-demultiplexed-seqs paired-end-demux.qza --o-representative-sequences rep-seqs.qza --o-table table.qza --o-denoising-stats DADA2Stats.qza
```
**注**：正向序列前有20 bp为测序引物，反向序列前端前15 bp为测序引物。

# 训练分类器

```
#下载qiime提供的silva数据库
## wget https://www.arb-silva.de/fileadmin/silva_databases/qiime/Silva_132_release.zip

#训练器需要的文件
含有rep_set，taxonomy目录

#V6V8区的引物
前引物序列926wF：AAACTYAAAKGAATTGRCGG
后引物序列1392R：ACGGGCGGTGWGTRC

#导入序列，指定格式为FeatureData[Sequence]，fasta格式
qiime tools import --type 'FeatureData[Sequence]' --input-path SILVA_138.1_SSURef_NR99_tax_silva_dna.fasta --output-path SSU_nr99.qza
#导入分类信息文件 
qiime tools import --type 'FeatureData[Taxonomy]' --input-format HeaderlessTSVTaxonomyFormat --input-path headers.txt --output-path ref-taxonomy.qza

#按照引物信息提取参考序列 28m19.399s
time qiime feature-classifier extract-reads --i-sequences SSU_nr99.qza --p-f-primer AAACTYAAAKGAATTGRCGG --p-r-primer ACGGGCGGTGWGTRC --p-trunc-len 0 --p-min-length 100 --p-max-length 550 --o-reads ref-seqs-Q2021.4_len550_AAACTYAAAKGAATTGRCGG-ACGGGCGGTGWGTRC.qza
#-p-trunc-len参数只能用于比对序列被剪裁成相同长度或更短的长度时，才需要剪裁参考序列。成功双端合并序列的长度通常是可变的。
#未在特定长度截断的单端读取的长度也可能是可变的。对于双端和未经修剪的单端读的物种分类，建议对在适当的引物位置提取但不修剪为等长的序列进行分类器训练。
#min-length参数在trim-left和trunc-len参数之后应用，在max-length之前应用，因此一定要设置适当的设置，以防止有效序列被过滤掉。
#训练分类器 318m32.905s
time qiime feature-classifier fit-classifier-naive-bayes --i-reference-reads ref-seqs-Q2021.4_len550_AAACTYAAAKGAATTGRCGG-ACGGGCGGTGWGTRC.qza --i-reference-taxonomy ref-taxonomy.qza --o-classifier classifier_for_V6V8_Q2021.4_len550_AAACTYAAAKGAATTGRCGG_ACGGGCGGTGWGTRC.qza

```

# 用SILVA数据库分类
real	24m36.136s
```
time qiime feature-classifier classify-sklearn --i-reads ./rep-seqs.qza --i-classifier ./classifier_for_V6V8_Q2021.4_len550_AAACTYAAAKGAATTGRCGG_ACGGGCGGTGWGTRC.qza --o-classification ./taxonomy.qza
```

# (可选)可视化物种注释为表，方便查看

```
qiime metadata tabulate \
  --m-input-file ./taxonomy.qza \
  --o-visualization ./taxonomy.qzv
```

# 过滤
## 过滤feature table
### 筛选最小频率为3，至少在1个样品中出现的特征
real	0m7.066s
```
time qiime feature-table filter-features \
  --i-table ./table.qza \
  --p-min-frequency 3 \
  --p-min-samples 1 \
  --o-filtered-table ./table_abund31.qza
```

### 去除Unassigned和未注释到Phylum的Features

```bash
qiime taxa filter-table \
  --i-table table_abund31.qza \
  --i-taxonomy taxonomy.qza \
  --p-include 'Bacteria;','Archaea;','Eukaryota;' \
  --p-exclude Unassigned \
  --o-filtered-table table_abund31-with-Phylum-no-Unassigned.qza
```
### 查看过滤后的特征表
real	0m13.321s
```
time qiime feature-table summarize \
  --i-table table_abund31-with-Phylum-no-Unassigned.qza \
  --o-visualization table_abund31-with-Phylum-no-Unassigned.qzv \
  --m-sample-metadata-file sample-metadata.tsv
```

## 过滤代表序列

### 去除Unassigned和未注释到Phylum的序列

```bash
qiime taxa filter-seqs \
  --i-sequences rep-seqs.qza \
  --i-taxonomy taxonomy.qza \
  --p-include 'Bacteria;','Archaea;','Eukaryota;' \
  --p-exclude Unassigned \
  --o-filtered-sequences rep-seqs-with-Phylum-no-Unassigned.qza
```
### (可选)将代表性序列制成表格。 对代表性序列进行制表将允许我们查看分配给标识符的序列，并以NCBI数据库的形式交互式地对序列进行比对查询

```bash
qiime feature-table tabulate-seqs \
  --i-data ./rep-seqs-with-Phylum-no-Unassigned.qza \
  --o-visualization ./rep-seqs-with-Phylum-no-Unassigned.qzv
```

# Alpha稀疏和深度选择
 real	1m7.621s
```
time qiime diversity alpha-rarefaction \
  --i-table ./table_abund31-with-Phylum-no-Unassigned.qza \
  --m-metadata-file ./sample-metadata.tsv \
  --o-visualization ./alpha_rarefaction_curves.qzv \
  --p-min-depth 10 \
  --p-max-depth 43266
```

# 构建多样性分析所需的进化树
## (有问题)将代表序列插入到参考树上
real	428m23.831s
```
time qiime fragment-insertion sepp \
  --i-representative-sequences ./rep-seqs-with-Phylum-no-Unassigned.qza \
  --i-reference-database sepp-refs-gg-13-8.qza \
  --o-tree ./tree.qza \
  --o-placements ./tree_placements.qza \
  --p-threads 20
  ```
  **注意**：并非所有的序列都出现在系统发育树上，具体原因不明。可用无参考序列的系统发育树进行替代。
  
## 仅构建代表序列系统发育树

### 分步法

#### 比对
```bash
qiime alignment mafft --i-sequences rep-seqs.qza --o-alignment aligned-rep-seqs.qza
```
#### 遮罩
```bash
qiime alignment mask --i-alignment aligned-rep-seqs.qza --o-masked-alignment masked-aligned-rep-seqs.qza
```
#### ……

### 管道
```bash
qiime phylogeny align-to-tree-mafft-fasttree --i-sequences rep-seqs-with-Phylum-no-Unassigned.qza --output-dir mafft-fasttree-output
```

# 多样性分析
real	0m25.782s
```
time qiime diversity core-metrics-phylogenetic \
  --i-table ./table_abund31-with-Phylum-no-Unassigned.qza \
  --i-phylogeny ./mafft-fasttree-output/rooted_tree.qza \
  --m-metadata-file ./sample-metadata.tsv \
  --p-sampling-depth 7702 \
  --output-dir ./core-metrics-results
```

# 使用过滤表在每个样本中构建分类法的交互式条形图
real	0m26.044s
```
time qiime taxa barplot \
  --i-table ./table_abund31-with-Phylum-no-Unassigned.qza \
  --i-taxonomy ./taxonomy.qza \
  --m-metadata-file ./sample-metadata.tsv \
  --o-visualization ./taxa_barplot.qzv
```

# 个性化分析

## 导出特征表,得到feature-table.biom
```
qiime tools export --input-path table_abund31-with-Phylum-no-Unassigned.qza --output-path exported-feature-table
```

## 将biom特征表转换为PCL格式文件

```
biom convert -i exported-feature-table/feature-table.biom -o exported-feature-table/feature-table.tsv --to-tsv --header-key taxonomy
```

## 绘制Alpha-diversity箱线图

### 安装软件
```r
library(devtools)
install_github("liaochenlanruo/amplicon")
```

### 文件准备
#### metaData：sample-metadata.tsv
#### FeatureTable：feature-table.tsv
首行为样本名，第一个单元格留空。
#### Tree：tree.nwk

### 脚本准备
#### compute_alpha.R
#### Plot_alpha.R

### 分析过程
#### 获取Alpha-diversity
```bash
Rscript compute_alpha.R
```
#### 绘图
```bash
Rscript Plot_alpha.R
```

## NMDS

### 文件准备
#### metaData：sample-metadata.tsv
#### FeatureTable：feature-table.tsv
首行为样本名，第一个单元格留空。

### 计算

```r
setwd("E:/Researches/Xiaqian/NGS/CleanData/ALL/细菌V6V8-1/Analysis_20210527")
library("vegan")

# 读取数据，一份otu.table文件和一份分组信息文件
otu <- read.table("feature-table.tsv",row.names=1, header=T,sep="\t",check.names=F)
design <- read.table("sample-metadata.tsv",header=T,sep="\t",row.names=1,check.names=F)

#数据调整为列名是OTU，行名是样本名
otu=t(otu)

#标准化，当然method有很多，可以通过?decostand查看其它method
vare.hel<-decostand(otu,method="hellinger")

#计算Bray-curtis距离矩阵
vare.dis <- vegdist(vare.hel,method="bray")

#使用NMDS的方法
vare.mds <- metaMDS(vare.hel,distance = "bray")

#首先提取前两轴坐标
point = scores(vare.mds)

#将分组文件和得分文件合并
index = merge(design, point,by="row.names",all=F)

#查看Stress值
#Stress值是反映模型合适程度的指标，NMDS会多次打乱数据计算Stress值，直到找到最合适的模型，也就是最低的Stress值；理想状况下，Stress值为0，一般Stress值低于0.1较为合理。
vare.mds
# Stress:     0.08814717 

#显着性检验；anosim本质是基于排名的算法更加适合NMDS
anosim.result<-anosim(vare.dis, design$Depth,permutations =999)
summary(anosim.result)

# ANOSIM statistic R: 0.2373 
#       Significance: 0.001
#tiff输出图形，适合大部分出版刊物，入门级别分辩率300,18*14的长宽；
tiff(file="beta_bray_NMDS.tif", res = 300, compression ="none", width=180,height=140,units= "mm")

#开始出图，将上面得到的三个指标在图中更换stress，R，p，不多说，代码如下：
library("ggplot2")

p = ggplot(index, aes(x=NMDS1, y=NMDS2, color=as.factor(Depth))) +
      geom_point(size=0) +
     scale_colour_manual(values = c("red","blue", "green", "#9900FF", "#336633", "#996600")) +
     labs(x=paste("NMDS1"), y=paste("NMDS2"), title="")

#置信区间当然要加上，有三种方式，线条类型也可以更改

p+stat_ellipse(type = "t", linetype = 2, level = 0.95, show.legend = TRUE)+
    annotate("text",x=-2.5,y=-1.25,parse=TRUE,size=4,label="'R: '*0.2373",family="serif",fontface="italic",colour="darkred",hjust = 0)+
    annotate("text",x=-2.5,y=-1.35,parse=TRUE,size=4,label="'p: '*0.001",family="serif",fontface="italic",colour="darkred",hjust = 0)+
    annotate("text",x=-2.5,y=-1.45,parse=TRUE,size=4,label="'Stress: '*0.088",family="serif",fontface="italic",colour="darkred",hjust = 0) + geom_point(aes(x=NMDS1, y=NMDS2, shape=Site), size=3)+
	scale_shape_manual(values = c(15, 19, 17, 18, 7, 8, 13))+
    theme_classic()

dev.off()
```

# trans_func class
```
# Identify microbial traits
# create object of trans_func
t2 <- trans_func$new(meco_dataset)
# mapping the taxonomy to the database
# the function can recognize prokaryotes or fungi automatically
t2$cal_spe_func()
# return t2$res_spe_func, 1 represent function exists, 0 represent no or cannot confirmed.
```

Tax4Fun requires a strict input file demand on the taxonomic information. To analyze the trimmed or changed OTU data in R with Tax4Fun, we provide a link to the Tax4Fun functional prediction.
```
t1 <- trans_func$new(meco_dataset)
# install Tax4Fun package and download SILVA123 ref data from  http://tax4fun.gobics.de/
wget https://github.com/bwemheu/Tax4Fun2/releases/download/1.1.5/Tax4Fun2_1.1.5.tar.gz
install.packages(pkgs = "Tax4Fun2_1.1.5.tar.gz", repos = NULL, source = TRUE)
# decompress SILVA123; provide path in folderReferenceData as you put
t1$cal_tax4fun(folderReferenceData = "./SILVA123")
# return two files: t1$tax4fun_KO: KO file; t1$tax4fun_path: pathway file.
# t1$tax4fun_KO$Tax4FunProfile[1:5, 1:2]
```
